File: Leetcode/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */
#include <algorithm>
#include <iostream>
#include <vector>
#include <unordered_map> 

using namespace std;
/*2021-09-22-11-18*/
// ? 思路解析：
/* 
 * 
 * 
 ! 
 ! https://www.acwing.com/activity/content/code/content/1809961/
 */

// @lc code=start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int ans = 0;
        unordered_map<int, int> hash;
        int i = 0;

        for (int j = 0; j < s.size(); j++) {
            hash[s[j]]++;
            while (hash[s[j]] > 1) hash[s[i++]]--;
            ans = max(ans, j - i + 1);
        }
        
        return ans;
    }
};
// @lc code=end

File: Leetcode/3.无重复字符的最长子串.go
/*
 * @lc app=leetcode.cn id=3 lang=golang
 *
 * [3] 无重复字符的最长子串
 */
package leetcode

/* --- 2022-04-06-16-21 --- */

// https://github.com/fengwei2002/Algorithm

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	// 大概思路就是维护一个动态的滑动窗口哈希表
	hash := make(map[byte]int, 0)
	ans := 0
	for l, r := 0, 0; r < len(s); r++ {
		hash[s[r]] += 1 
		for {
			cnt, _ := hash[s[r]]
			if cnt > 1 {
				hash[s[l]] -= 1
				l += 1
			} else {
				break
			}
		}
		ans = max(ans, r-l+1)
	}
	return ans
}

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}
// @lc code=end
